core/commands/ProjectHierarchyGetCommand.php

```diff
@@ -0,0 +1,156 @@
+
+# MantisBT is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# MantisBT is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with MantisBT.  If not, see <http://www.gnu.org/licenses/>.
+
+require_api( 'constant_inc.php' );
+require_api( 'config_api.php' );
+require_api( 'helper_api.php' );
+require_api( 'authentication_api.php' );
+require_api( 'user_api.php' );
+require_api( 'project_api.php' );
+require_api( 'project_hierarchy_api.php' );
+
+use Mantis\Exceptions\ClientException;
+
+/**
+ * A command that gets project subprojects from the project hierarchy.
+ */
+class ProjectHierarchyGetCommand extends Command {
+
+	/**
+	 * @var integer
+	 */
+	private $project_id;
+
+	/**
+	 * @var boolean
+	 */
+	private $all_descendents;
+
+	/**
+	 * @var boolean
+	 */
+	private $include_disabled;
+
+	/**
+	 * return an array of user accessible subprojects of a project (recurse if requested)
+	 * @param integer $p_project_id	A valid project identifier.
+	 * @param boolean $p_all_descendents Include only childs or all descendents.
+	 * @param boolean $p_include_disabled Include disabled projects in the resulting array.
+	 * @param array $p_subprojects The current array of subprojects.
+	 * @return array
+	 */
+	private function get_subprojects( $p_project_id, $p_all_descendents, $p_include_disabled, $p_subprojects  = array() ) {
+		$t_subprojects = $p_subprojects;
+		$t_subproject_ids = user_get_accessible_subprojects(
+			auth_get_current_user_id(), $p_project_id, $p_include_disabled );
+		foreach( $t_subproject_ids as $t_subproject_id ) {
+			$t_subprojects[] = array(
+				'id' => $t_subproject_id,
+				'name' => project_get_name( $t_subproject_id ),
+				'parent_id' => $p_project_id,
+				'inherit_parent' => project_hierarchy_inherit_parent(
+					 $t_subproject_id, $p_project_id, $p_include_disabled )
+			);
+
+			if( true === $p_all_descendents ) {
+				$t_subprojects += $this->get_subprojects(
+					$t_subproject_id, $p_all_descendents, $p_include_disabled, $t_subprojects );
+			}
+		}
+
+		return $t_subprojects;
+	}
+
+	/**
+	 * Constructor
+	 *
+	 * $p_data['query'] is expected to contain:
+	 * - project_id (integer)
+	 *
+	 * $p_data['payload'] is expected to contain:
+	 * - all_descendents (boolean)
+	 * - include_disabled (boolean)
+	 *
+	 * @param array $p_data The command data.
+	 */
+	function __construct( array $p_data ) {
+		parent::__construct( $p_data );
+	}
+
+	/**
+	 * Validate the data.
+	 */
+	function validate() {		
+		if ( config_get( 'subprojects_enabled' ) == OFF ) {
```

Comment:
This might be a semantic nitpick, but `subprojects_enabled` doesn't tell if there are subprojects at all. Perhaps created and never cleaned up before subprojects were disabled. I'd rather stick with erroring out than returning anything not matching current state.

Reply:
Perhaps an alternative might be to omit the check for `subprojects_enabled` and just return anything in there. This would be most likely `array()` anyway.